infrastructure/src/test/java/org/corfudb/infrastructure/management/failuredetector/EpochHandlerTest.java

```diff
@@ -0,0 +1,181 @@
+    }
+
+    @Test
+    void fetchLatestLayoutConnectionError() {
+        CompletableFuture<Layout> err = new CompletableFuture<>();
+        err.completeExceptionally(new WrongEpochException(123));
+
+        CompletableFuture<Layout> err2 = new CompletableFuture<>();
+        err2.completeExceptionally(new WrongEpochException(321));
+
+        Map<String, CompletableFuture<Layout>> futureLayouts = new HashMap<>();
+        futureLayouts.put(NodeNames.A, err);
+        futureLayouts.put(NodeNames.B, err2);
+
+        Optional<Layout> latestLayout = buildHandler()
+                .fetchLatestLayout(futureLayouts)
+                .join();
+
+        assertFalse(latestLayout.isPresent());
+    }
+
+    @Test
+    void fetchLatestLayoutOneServer() {
+        CompletableFuture<Layout> cf0 = CompletableFuture.completedFuture(buildSimpleLayout());
+
+        HashMap<String, CompletableFuture<Layout>> futureLayouts = new HashMap<>();
+        futureLayouts.put(NodeNames.A, cf0);
+
+        Optional<Layout> latestLayout = buildHandler()
+                .fetchLatestLayout(futureLayouts)
+                .join();
+
+        assertTrue(latestLayout.isPresent());
+    }
+
+    @Test
+    void fetchLatestLayoutThreeNodesOneDown() {
+        final int highestEpoch = 33;
+
+        CompletableFuture<Layout> cf0 = CompletableFuture.completedFuture(buildSimpleLayout());
+        CompletableFuture<Layout> cf1 = CompletableFuture.completedFuture(buildSimpleLayout(highestEpoch));
+        CompletableFuture<Layout> cf2 = new CompletableFuture<>();
+        cf2.completeExceptionally(new WrongEpochException(123));
+
+        HashMap<String, CompletableFuture<Layout>> futureLayouts = new HashMap<>();
+        futureLayouts.put(NodeNames.A, cf0);
+        futureLayouts.put(NodeNames.B, cf1);
+        futureLayouts.put(NodeNames.C, cf2);
+
+        Optional<Layout> latestLayout = buildHandler()
+                .fetchLatestLayout(futureLayouts)
+                .join();
+
+        assertTrue(latestLayout.isPresent());
+        assertEquals(highestEpoch, latestLayout.get().getEpoch());
+    }
+
+    @Test
+    void updateTrailingLayoutServersCommitLayoutExecution() {
+        EpochHandler handlerSpy = Mockito.spy(buildHandler());
+
+        HashMap<String, CompletableFuture<Layout>> requests = new HashMap<>();
+        CompletableFuture<Layout> request1 = new CompletableFuture<>();
+        request1.complete(buildSimpleLayout(2));
+        requests.put(NodeNames.A, request1);
+
+        CompletableFuture<Boolean> asyncCommit = CompletableFuture.completedFuture(true);
+        doReturn(asyncCommit).when(handlerSpy).commitLayout(isA(Layout.class), anyString());
+
+        Layout layout1 = buildSimpleLayout();
+        handlerSpy.updateTrailingLayoutServers(layout1, requests);
+
+        verify(handlerSpy, times(1)).commitLayout(any(Layout.class), anyString());
+    }
+
+    @Test
+    void updateTrailingLayoutServersFailedConnections() {
```

Comment:
Zack, I didn't get your question, could you please explain?

I can say that we execute `commitLayout`, which must update the layout (because of Paxos) or fail; there is nothing in between

Reply:
In `updateTrailingLayoutServersCommitLayoutExecution`, there is only one node in the request map, and we expect to update this node's `Layout`. Likewise in this test, we have three nodes in the request map, and again expect to update all of them. We don't have a test where, when we iterate over the layout servers in the map, only some of them need to have their Layouts updated (i.e. for some nodes in the map we have `LayoutCommit.NO_NEED_UPDATE`, and for others, we have `LayoutCommit.NEED_UPDATE`).

Reply:
Got it, that will be a good test, I'll add that into part 3